File: python-to-go-upskill/Day1a/code/Stack-and-Queue-Reversal.py
"""
Stack and queue reversal examples for new learners.

This file shows two common reversal techniques in Python:
- reversing a simple list used like a stack
- reversing a deque used like a queue

A stack is a last-in-first-out (LIFO) structure. The easiest reversal
for a stack-like list is using slicing with `[::-1]`.

A queue is a first-in-first-out (FIFO) structure. When using a deque,
calling `reverse()` changes the order of elements in place.
"""

# Stack reversal example
stack = [1, 2, 3, 4]
reversed_stack = stack[::-1]
print("Stack input :", stack)
print("Stack output:", reversed_stack)
# Output: [4, 3, 2, 1]

# Queue reversal example using deque
from collections import deque
q = deque([1, 2, 3, 4])
q.reverse()
print("Queue input :", deque([1, 2, 3, 4]))
print("Queue output:", q)
# Output: deque([4, 3, 2, 1])

# Complexity:
# - Stack reversal with slicing: O(n) time, O(n) space because a new list is created.
# - Deque reversal with q.reverse(): O(n) time, O(1) extra space because it reverses in place.
#
# Useful info:
# - Lists are often used like stacks in Python because append/pop work well.
# - Deques are better for queues because they are efficient at both ends.
# - Use `stack[::-1]` for a quick reversed copy; use `reversed(stack)` if you only need to iterate backwards.
# - Use `deque.reverse()` when you want to change the queue order without creating a new object.

File: python-to-go-upskill/Day1a/code/Stack-and-Queue-Reversal.go
// Stack and queue reversal with slices in Go.
//
// This program shows how to reverse a sequence of values using a slice.
// In Go, slices are commonly used for both stack-like and queue-like behavior.
package main

import "fmt"

func reverseSeq(s []int) []int {
	for i, j := 0, len(s)-1; i < j; i, j = i+1, j-1 {
		s[i], s[j] = s[j], s[i]
	}
	return s
}

func main() {
	inputSlice := []int{1, 2, 3, 4, 5}
	result := reverseSeq(inputSlice)

	fmt.Println("Input  :", inputSlice)
	fmt.Println("Output :", result)
	fmt.Printf("input %v output %v\n", inputSlice, result)

	// Complexity:
	// - Time complexity: O(n) because each element is swapped at most once.
	// - Space complexity: O(1) additional space because the reversal happens in place.
	//
	// Useful info:
	// - Slices are flexible, and reversing a slice modifies the underlying data.
	// - If you need to preserve the original slice, make a copy first:
	//   `copySlice := append([]int(nil), inputSlice...)`
	// - Use slices for both stack and queue patterns, but choose a clear API
	//   for your intended behavior.
}

// Input  : [1,2,3,4,5]
// Output : [5,4,3,2,1]
